cardano-cli/src/Cardano/CLI/Slot.hs

```diff
@@ -0,0 +1,329 @@
+import           Data.Kind (Type)
+import           Data.Maybe (fromMaybe)
+import           Data.Ratio ((%))
+import           Data.Time.Clock (NominalDiffTime, UTCTime, getCurrentTime)
+import           Data.Word (Word32)
+import           GHC.Generics (Generic)
+import           GHC.Stack (HasCallStack)
+import           GHC.TypeLits (Symbol)
+import           Ouroboros.Consensus.BlockchainTime.WallClock.Types (RelativeTime (..),
+                   SystemStart (..), diffRelTime)
+import           Ouroboros.Consensus.HardFork.History.Qry (Interpreter, PastHorizonException (..),
+                   mkInterpreter, slotToWallclock)
+import           Ouroboros.Consensus.HardFork.History.Summary (neverForksSummary)
+import           Quiet (Quiet (..))
+
+import qualified Ouroboros.Consensus.BlockchainTime.WallClock.Types as Cardano
+import qualified Ouroboros.Consensus.HardFork.History.Qry as HF
+
+data TimeInterpreterLog = MsgInterpreterPastHorizon
+  (Maybe String) -- ^ Reason for why the failure should be impossible
+  StartTime
+  PastHorizonException
+  deriving (Show)
+
+-- | Number of slots in a single epoch
+newtype EpochLength = EpochLength { unEpochLength :: Word32 }
+    deriving (Show, Eq, Generic)
+
+-- | Duration of a single slot.
+newtype SlotLength = SlotLength { unSlotLength :: NominalDiffTime }
+    deriving (Show, Eq, Generic)
+
+-- | Blockchain start time
+newtype StartTime = StartTime UTCTime
+    deriving (Show, Eq, Ord, Generic)
+
+-- | A @TimeInterpreter@ is a way for the wallet to run things of type @Qry a@,
+-- with a system start time as context.
+data TimeInterpreter m = forall eras. TimeInterpreter
+  { interpreter :: m (Interpreter eras)
+  , blockchainStartTime :: StartTime
+  }
+
+newtype Quantity (unit :: Symbol) a = Quantity { getQuantity :: a }
+  deriving stock (Generic, Show, Eq, Ord)
+  deriving newtype (Bounded, Enum, Hashable)
+  deriving newtype NFData
+
+newtype Hash (tag :: Symbol) = Hash { getHash :: ByteString }
+  deriving stock (Generic, Eq, Ord)
+  deriving (Read, Show) via (Quiet (Hash tag))
+  deriving anyclass (NFData, Hashable)
+
+newtype SyncTolerance = SyncTolerance NominalDiffTime
+  deriving stock (Generic, Eq, Show)
+
+data SyncProgress
+  = Ready
+  | Syncing !(Quantity "percent" Percentage)
+  | NotResponding
+  deriving (Generic, Eq, Show)
+
+newtype Percentage = Percentage
+  { getPercentage :: Rational
+  }
+  deriving stock (Generic, Show, Eq, Ord)
+
+-- | A query for time, slot and epoch conversions. Can be interpreted using
+-- @interpretQuery@.
+--
+-- == Differences to the underlying consensus 'Ouroboros.Consensus.HardFork.History.Qry.Qry'
+--
+-- @HF.Qry@ can only be interpreted in a
+-- single era. If you have
+--
+-- @
+--     q1 = epochOf someSlotInByron
+--     q2 = epochOf someSlotInShelley
+-- @
+--
+-- @HF.interpretQuery@ could interpret both individually, but
+--
+-- @
+--    q3 = q1 >> q2
+-- @
+--
+-- would fail.
+--
+-- This wrapper exists to fix this.
```

Comment:
Do you still need this extra layer of `Qry`? In particular, I only see `interpretQuery ti $ slotToRelTime tipSlotNo` used here, and that should only require a single era. (I do see the PR is marked as Draft, so maybe there's more coming.)

Reply:
There's not more coming.  This code was adapted from what I found in `cardano-wallet`, so I preserved the structure and am looking to untangle and simplify it.  If I can get away with single era for this query that would be great!